google/cloud/storage/list_buckets_reader.cc

```diff
@@ -0,0 +1,96 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/list_buckets_reader.h"
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+// ListBucketsReader::iterator must satisfy the requirements of an
+// InputIterator.
+static_assert(
+    std::is_base_of<std::iterator<std::input_iterator_tag, BucketMetadata>,
+                    ListBucketsReader::iterator>::value,
+    "ListBucketsReader::iterator must be an InputIterator");
+static_assert(std::is_copy_constructible<ListBucketsReader::iterator>::value,
+              "ListBucketsReader::iterator must be CopyConstructible");
+static_assert(std::is_move_constructible<ListBucketsReader::iterator>::value,
+              "ListBucketsReader::iterator must be MoveConstructible");
+static_assert(std::is_copy_assignable<ListBucketsReader::iterator>::value,
+              "ListBucketsReader::iterator must be CopyAssignable");
+static_assert(std::is_move_assignable<ListBucketsReader::iterator>::value,
+              "ListBucketsReader::iterator must be MoveAssignable");
+static_assert(std::is_destructible<ListBucketsReader::iterator>::value,
+              "ListBucketsReader::iterator must be Destructible");
+static_assert(
+    std::is_convertible<decltype(*std::declval<ListBucketsReader::iterator>()),
+                        ListBucketsReader::iterator::value_type>::value,
+    "*it when it is of ListBucketsReader::iterator type must be convertible to "
+    "ListBucketsReader::iterator::value_type>");
+static_assert(
+    std::is_same<decltype(++std::declval<ListBucketsReader::iterator>()),
+                 ListBucketsReader::iterator&>::value,
+    "++it when it is of ListBucketsReader::iterator type must be a "
+    "ListBucketsReader::iterator &>");
+
+ListBucketsIterator::ListBucketsIterator(
+    ListBucketsReader* owner,
+    google::cloud::internal::optional<BucketMetadata> value)
+    : owner_(owner), value_(std::move(value)) {
+  if (not value_) {
+    // This iterator was initialized by begin() on an empty list, turn it into
+    // an end() iterator.
+    owner_ = nullptr;
+  }
+}
+
+ListBucketsIterator& ListBucketsIterator::operator++() {
+  value_ = owner_->GetNext();
+  if (not value_) {
+    owner_ = nullptr;
+  }
+  return *this;
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+ListBucketsReader::iterator ListBucketsReader::begin() {
+  return iterator(this, GetNext());
+}
+
+google::cloud::internal::optional<BucketMetadata> ListBucketsReader::GetNext() {
+  if (current_buckets_.end() == current_) {
```

Comment:
I think little comments above each of these if statements would be helpful to cover which edge cases we're accounting for, e.g. "If we're at the end of the current page of items", "we've iterated over all items in all pages", "the next page of items was empty, so we've iterated over all items in all pages", etc.

Reply:
Done.